[bump minor] Cross exposure

corentinravoux opened the pull request:
Large modifications of the pk1d routines, and some modifications at the delta extraction. The main point of the PR is to implement an estimator of the one-dimensional power spectrum based on cross-correlations between different exposures of the same quasar. This PR can be extended to the computation of P3D (P_cross in particular).

For P3D: Just use picca_Pk1d.py, and the new individual pk1d file contains the deltas' Fourier transforms.
For P1D cross-exposures: first use picca_delta_extraction.py with the options (use non-coadded spectra = True and keep single exposures = True) in [data], and calculate the Fourier transforms of the deltas with picca_Pk1d.py. Then, I've added a picca_Pk1D_cross_exposure.py script that estimates individual P1D cross-exposures.

The hope of this PR is to have a proper P1D cross-exposure estimator, and to start the process of implementing P3D estimators directly inside picca.

Waelthus commented:
merged master back into here to enable the changes from v9, please double check... removed the "unblind" option. Not sure if that was from the starting point of this PR or one of your changes though...

corentinravoux commented:
Yes this seems ok for me, I did not add or removed the unblind strategy. And the only conflict I saw was only on the addition of this option.

corentinravoux commented:
@Waelthus @iprafols  I would like to have your opinion especially on the modifications of delta extraction on which I worked on. Since it does not only impact the P1D.

iprafols commented:
I did not check the pk1d changes in a lot of detail, let me know when you are ready for me to do so. 

Also, on the changes of delta_extraction, things are starting to get more and more complicated. Maybe we should consider splitting the class `DesiData` into `DesiData3d` and `DesiData1d`, and have the `DesiData1d` inherit from the `DesiData3d`. I'm not saying we should do this for this PR but I do think we should discuss this at the next telecon

corentinravoux commented on bin/picca_Pk1D_cross_exposure.py:
```diff
@@ -0,0 +1,158 @@
+from functools import partial
+
+
+def treat_pk_file(out_dir, filename):
```
Docstring added

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -106,6 +111,7 @@ def __init__(self, config):
 
         # load variables from config
         self.use_non_coadded_spectra = None
+        self.keep_single_exposures = None
```
Corrected

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -145,6 +151,13 @@ def __parse_config(self, config):
             raise DataError(
                 "Missing argument 'use non-coadded spectra' required by DesiData"
             )
+        self.keep_single_exposures = config.getboolean(
```
Corrected

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,80 +332,164 @@ def format_data(self,
-                    ivar = ivar.sum(axis=0)
-                    flux = ivar_coadded_flux / ivar
+                    if self.keep_single_exposures:
+                        flux = spec['FLUX'][w_t].copy()
```
Yes indeed, corrected

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,80 +332,164 @@ def format_data(self,
+                            num_data)
         return forests_by_targetid, num_data
 
+    def update_forest_dictionary(self,
```
Yes this is correct. Since I am doing that in two different manners for cross-exposure and regular, I choose to not repeat it.

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,80 +332,164 @@ def format_data(self,
+                    if len(flux.shape) != 2:
+                        flux , ivar = np.array([flux]), np.array([ivar])
+                    for i, flux_i in enumerate(flux):
+                        num_data = self.update_forest_dictionary(
```
yes, this is done now

corentinravoux commented:
Yes we could separate the reading between 3D and 1D. However, I am wondering if the changes done on the 3D and 1D analysis at this stage might not be beneficial for the other analysis? Maybe the cross-exposure could be used for the 3D one day ?

Waelthus commented on py/picca/pk1d/postproc_pk1d.py:
```diff
@@ -221,6 +227,24 @@ def compute_mean_pk1d(
     if "sub_forest_id" in p1d_table_cols:
         p1d_table_cols.remove("sub_forest_id")
 
+    if "Delta_k_real" in p1d_table_cols:
```
Why are those cols removed (esp given that you just explicitely added them to the readin part above)? is it because some iterations further down go over all existing cols? In which case maybe a check there would be more compact (and more clearly tell why this is done) as in `if "Delta_" in col: continue`

corentinravoux replied:
The explicit part above is just for renaming. The Table reads all the file.
And yes there is an iteration further down that go over all columns. We do not want to average Detla_k for nothing.
This is quite equivalent no ? I can change that though

corentinravoux replied:
After checking more carefully the code, your change would modify a lot of things, since the statistics array are first create in a list and then computed in another. 
I will simplify the way I wrote it.

Waelthus replied:
I was mostly confused because these fields have not been used anywhere so-far, yet they are first renamed further up (which I now noticed works from uppercase as probably stored in the FITS to the actual case used for those variables) and then removed...
I guess explicitely writing this down makes it clearer. Else a comment in the file such as "#remove columns we don't need averages of" might be helpful to clarify...

Waelthus commented on py/picca/pk1d/postproc_pk1d.py:
```diff
@@ -73,6 +73,12 @@ def read_pk1d(filename, kbin_edges, snrcut=None, zbins_snrcut=None):
                 "Pk_diff",
                 "cor_reso",
                 "Pk_noise_miss",
+                "Delta_k_real",
```
similarly, use complex numbers to simplify things

Waelthus commented on py/picca/tests/test_2_pk1d.py:
```diff
@@ -69,7 +69,7 @@ def test_Pk1D_raw(self):
         if self._test:
             path1 = self._masterFiles + "/test_Pk1D/Pk1D_raw.fits.gz"
-            path2 = self._branchFiles + "/Products/Pk1D_raw/Pk1D-0.fits.gz"
+            path2 = self._branchFiles + "/Products/Pk1D_raw/Pk1D-272.fits.gz"
```
those numbers are now propagated from the delta files, right? So they are deterministic?

corentinravoux replied:
yes, I choose to save the healpix number instead of the index number which does not make sense. It will be usefull for P3D stuff

Waelthus reviewed (commented):
Just leaving some comments for the moment:
- generally would be prettier if at some parts actual complex numbers were used
- maybe change the TARGETID_OTHERID to something that does not change TARGETID, but adds another general descriminator

Waelthus commented on bin/picca_Pk1D_cross_exposure.py:
```diff
@@ -0,0 +1,230 @@
+                [fft_delta_list[i].mean_snr for i in index]
+            )
+            mean_reso = np.mean([fft_delta_list[i].mean_reso for i in index])
+            fft_delta_real = np.array([fft_delta_list[i].fft_delta_real for i in index])
```
also here, I think it would be easier to store everything in complex numbers

corentinravoux commented:
Sorry for the long time of this PR, but I did not forgot your comments @Waelthus ! I will first try to solve the DLA masking issue, and I will come back to address those

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -145,6 +151,13 @@ def __parse_config(self, config):
+            raise DataError(
+                "Missing argument 'keep single exposures' required by DesiData"
+            )
+
```
Yes I change this test. Do you mean that we should add a test mode for cross-exposure ? That is indeed a good idea. For now since the cross-exp is only preliminary, I suggest to postpone the test to the next PR

Waelthus replied:
I guess we should try to have multiple tests:
1) on the delta_extraction side tests for the different arguments allowed in the config now would be good so that future updates won't accidentally break this mode
2) a test using the outputs of delta_extraction for a well defined small dataset in the required settings as an input to the cross_pk routines and running those could be good to at least have a consistency check when putting more changes into the code

corentinravoux commented on bin/picca_Pk1D.py:
```diff
@@ -14,54 +14,23 @@
 from multiprocessing import Pool
 
 
-def check_linear_binning(delta):
```
yes I moved your function to the compute_pk1d part. Later, I think we shoud have minimal number of lines in the script, this is not the case here.

corentinravoux commented on bin/picca_Pk1D.py:
```diff
@@ -203,59 +172,62 @@ def process_all_files(index_file_args):
+            k, fft_delta, pk_raw = compute_pk_raw(pixel_step,
+                                                  delta_new,
+                                                  linear_binning=linear_binning)
+            fft_delta_real = fft_delta.real
```
I hesitate between complex and separating imaginary and real part.
My choice mainly comes from the fact that FITS is not able to store complex, from what I saw on the web.
Do you still think we should reconstruct the complex each time we read ?

corentinravoux commented on bin/picca_Pk1D_cross_exposure.py:
```diff
@@ -0,0 +1,230 @@
+    with fitsio.FITS(filename, "r") as hdus:
+        for _, hdu in enumerate(hdus[1:]):
+            fft_delta = Pk1D.from_fitsio(hdu)
+            if len(fft_delta.los_id.split("_")) < 2:
```
Here CHUNKID comes from an older PR: it is number of a sub-forest after cutting in three sub-forests.
I think you are mistaking it with an exposure id. And yes, in that case this will be changed because the DLA masking is not done properly when I name the forests TARGETID_EXPOSUREID.
Yes the best solution is to create a new field which is EXPOSUREID, but it requires way more work since, for the tile format for example, there are some exposures that are located on different tiles. 
I used the TARGETID_EXPOSUREID as a provisory solution, I will change that

Waelthus replied:
ah, sorry that's right. I was so much into multi-exp thinking that I forgot about the chunking after...
I guess one could always use either EXPID (in case of using non-coadded spectra files as input) or TILEID (in case of using coadds from tile-based structures) as a descriminator and store them in a unified name to the outputs of delta_extraction, then read that unified name here...

corentinravoux commented on bin/picca_Pk1D_cross_exposure.py:
```diff
@@ -0,0 +1,230 @@
+            pk_diff = np.zeros_like(pk_raw_cross_exposure)
+
+            correction_reso = np.mean(
+                [fft_delta_list[i].correction_reso for i in index], axis=0
```
I have made changes on this, now I am using delta_k / cor_reso(exposure) and applying the cross-exposure estimate. The "correction_reso" is only here to check, not for the computation of P1D

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,78 +334,162 @@ def format_data(self,
+                # To avoid coadding exposures, we store the forest with
+                # targetid_i (i = exposure index) instead of targetid
+                # The variable forests_by_targetid is modified by update_forest_dictionary.
+                if self.use_non_coadded_spectra & self.keep_single_exposures:
```
That is a good idea, I have not tested to much the tile mode, but I can add it to my todo list

Waelthus replied:
yes, I think we didn't use the tile mode too much anymore after fuji reductions/catalogs were made, but in principle for some of the cross_exposure tests they might be more useful (as then each remaining coadded exp would be still ~1200s Teff)...

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,78 +334,162 @@ def format_data(self,
+                    for i, flux_i in enumerate(flux):
+                        forests_by_targetid, num_data = self.update_forest_dictionary(
+                                forests_by_targetid,
+                                f"{targetid}_{i}",
```
Yes I totally agree with you. I started cross-exposure with no big hope. Now that we know that it work pretty well and that it can be used for other studies, I want to remove this hacky stuff.
But, it requires a lot of work indeed

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desi_data.py:
```diff
@@ -311,80 +332,164 @@ def format_data(self,
+                                spec,
+                                row,
+                                flux_i,
+                                ivar[i],
```
yes I will change that, thanks

corentinravoux commented on py/picca/delta_extraction/data_catalogues/desisim_mocks.py:
```diff
@@ -49,6 +49,10 @@ def __init__(self, config):
             self.logger.warning(
                 'the "use_non_coadded_spectra" option was set, '
                 'but has no effect on Mocks, will proceed as normal')
+        if self.keep_single_exposures:
```
Yes for now, all the mocks are coadded. And we treat the pixsim mocks with healpix or tiles modes, so I do not know if using it on mocks is really useful ?

Waelthus replied:
I guess one should do tests like these on mocks just to figure out what to expect. I think the error messages might be misleading when we'd actually do so (but are perfectly fine atm), that's all...

corentinravoux commented:
@Waelthus I checked your comment: I agree on the complex part, and I will implement the small changes you suggest.
I am currently working (when I have time) on the proper separation between TARGETID and EXPOSUREID, it will be usefull for DLA masking and make the code cleaner.

corentinravoux commented on bin/picca_Pk1D_cross_exposure.py:
```diff
@@ -50,33 +62,74 @@ def treat_pk_file(out_dir, filename):
-            pk_noise_cross_exposure = compute_pk_cross_exposure(
-                fft_delta_noise_real, fft_delta_noise_imag
+            pk_noise_cross_exposure = (
+                2 * pk_noise_cross_exposure_cross - pk_noise_cross_exposure_auto
```
This decomposition is not simple. I made all the math on a notebook if you want.
I did not used the real/imag decomposition here. And pk_noise_cross and pk_noise_auto are not complex, they are cross-exposure estimators.
In any case, It should all be equal to zero, and I verified it. It is a correlation between Gausian random realizations

Waelthus replied:
Yes, when everything is uncorrelated things should come out to be 0. But if this is written as complex numbers (without the explicit real/imag split) I guess the `.real` might still be needed.

corentinravoux commented on py/picca/pk1d/postproc_pk1d.py:
```diff
@@ -737,6 +761,8 @@ def compute_average_pk_redshift(
                 standard_dev[~np.isnan(standard_dev)],
                 snr_bins[~np.isnan(standard_dev)],
             )
+            if len(standard_dev) == 0:
```
I had one case when the standard_dev was only one np.nan, it made the whole thing crash. This is fixing this case.

Waelthus replied:
maybe checking for `np.all(np.isnan(standard_dev))` before computing the conversion factor would be more explicit, but fine either way, maybe add a comment if keeping it as is...

corentinravoux commented on bin/picca_plot_pk1d.py:
```diff
@@ -1,176 +0,0 @@
-#!/usr/bin/env python
```
It is not used anymore. and yes postprocessing is doing most of the work now

Waelthus replied:
ok, then the removing is totally reasonable

corentinravoux commented:
This is a working version: the spectra are read independently for different exposure, the masks are applied to individual deltas, and the continuum fitting is performed on the coadd of the exposures for each quasar.

Waelthus commented:
Limits for coveralls are set to fail when coverage decreases by > 1%. So this is the expected behaviour given a 1.2% change in coverage.
[The coveralls output](https://coveralls.io/builds/66830898) shows that the most strongly affected files are `py/picca/pk1d/compute_pk1d.py` (most lines modified, but coverage actually rose) and `py/picca/delta_extraction/data_catalogues/desi_data.py` (mostly the `verify_exposures_shape` function is completely untested, but also some of the branchings due to the new options).
So if you could add a test for cross-exposure configs in delta extraction, coveralls should be fine again. An actual end-to-end test of the cross-exposure pk1d calculation and post-processing is not needed for coveralls to pass, but would be useful (at the very latest when the cross exposure calculation actually gives a trustable result).

corentinravoux commented:
Ok thanks! My plan to finish this PR which starts to be long:
- Implement cross-exposure test (minimal or end-to-end)
- Store the fft deltas in complex
- Test on iron

Waelthus commented:
in case the current analysis routines are expected to already be able to do things properly (sry, didn't check), it might be worth changing the order (also so that the more interesting parts come first), i.e. 
- testing the current on iron (or any smaller data set)
- implementing a test
- changing the routines to be in complex, where now you have the test to compare if results would change

corentinravoux commented:
I implemented the coaddition of exposures before the delta extraction.
After a test on Y1 data, and even with applying multprocessing, the coaddition of a large number of data take to much time (Wall time on one perlmutter node). I added an option to not perform this step ("coadd before delta extraction").
When it is True, the delta extraction is done on the coadd and applied to individual exposures (the proper way, but the one which takes too much time on Y1)
When it is False, the delta extraction is done on one of the exposure and applied to all exposure. I am testing this scheme now on Y1.

The conclusion associated to this comment is that: **the coadd by rebinning routine does not seem to be optimized, it can be something that we should look into**

corentinravoux commented:
I changed my mind: I think we should try to avoid coadding with picca as much as possible. I will modify the cross-exposure so that the delta extraction is done independently for every exposures.

Waelthus commented:
One question regarding the coadding: shouldn't one do co-adding for continuum fitting always? Or at the very least make sure that the linear quasar continuum rescaling fit to each object is the same for different spectra with same TARGETID? Else the same spectrum might have completely different coefficients for the different exposures (which I'd guess would at the very least lead to additional continuum fitting noise). But not sure why you want to avoid co-adding in delta extraction

corentinravoux commented:
> One question regarding the coadding: shouldn't one do co-adding for continuum fitting always? Or at the very least make sure that the linear quasar continuum rescaling fit to each object is the same for different spectra with same TARGETID? Else the same spectrum might have completely different coefficients for the different exposures (which I'd guess would at the very least lead to additional continuum fitting noise). But not sure why you want to avoid co-adding in delta extraction

I will summarize the way we can do this in cross exposure more, as well as what I found for now:
- Continuum fitting on only one exposure for a quasar (wrong): This was the first implementation, and the simplest one. For a given target id, all the exposures had a continuum fitting, but only the last exposure considered was saved by picca. 
- Continuum fitting on picca coadd: To perform this (implemented at commit 9e89216aa467838cebce6642a41d59839b66434b), picca needs to rebin all spectra individually, and separately coadd them by rebinning. It works well on a small sample. The problem is for iron, I am hitting Walltime, even with a multiprocessing of the rebinning step.
- Continuum fitting on coadd: We can read both pipeline coadd and exposures by picca but it requires to entirely change the logic of picca. And the reading time is already quite long.
- Continuum fitting on individual exposure: This is the option I was chosen. As you mentioned this might not be optimal but this is achievable

corentinravoux commented:
I answered your comment before you edited it. No for now I am doing it for individual exposures separately, not the best one. This is a good idea. I can test both maybe.

Waelthus commented:
The issue is that it might take a lot of changes in how I/O or cont fitting is done. Maybe it would be useful to have a look at the different continuum fitting coefficients in the run you already have for the same objects to see how much of an error is done when doing it the current way...

corentinravoux commented:
@iprafols @Waelthus For the coverage failure, and thus the tests of the delta extraction, I do not understand how they work. Do I need to add a test_desi_healpix_read_file for cross-exposure inside the delta_extraction/tests/data_tests.py ? Or a test_cross_exposure inside the delta_extraction/tests/expected_flux_tests.py file ?
Those tests seems very unitary, is there an end-to-end delta test located in another part of picca that I did not see ?

corentinravoux commented:
Additionally, with cross-exposure I need to read "spectra-" files. They are available for the tile mode but not for the healpix mode. It can be an issue for the tests I want to implement ?

Waelthus commented:
Ok, I think what needs to be done to make the coverage go up is having tests analogous to add tests like [the tile tests in data_tests](https://github.com/igmhub/picca/blob/eecadaa92d1acdf8a78f55814d74d5d41d1f351f/py/picca/tests/delta_extraction/data_tests.py#L1163) with config settings that actually use the `keep_single_exposures` and `PK 1D` modes in a new function `test_desi_tile_no_coadd` or similar. I guess we'd need that also for [the equivalent routines for healpix](https://github.com/igmhub/picca/blob/eecadaa92d1acdf8a78f55814d74d5d41d1f351f/py/picca/tests/delta_extraction/data_tests.py#L947) (and having the equivalent spectra files available for the healpix would be needed there, but note that the tile test should already go through most of the additional code in `desi_data.py`, so maybe not 100% needed to do the healpix test atm).
Those should test if the data is read in correctly (including functions like `verify_exposures_shape`) and if you have the right number of spectra available afterwards etc...

Some functions for coadding, e.g. [this](https://github.com/igmhub/picca/blob/0f2b7b19b75f9ef61db3d180bb1c535362f7e8f0/py/picca/delta_extraction/data.py#L679C9-L679C31) are not used anywhere in the code yet, so either should be made usable via some option or be removed, I guess at `survey.py` as an alternative to `select_best_exposure`? That option needs to then have tests as well.

Things like select_best_exposures would need to have the equivalent test similar to [here](https://github.com/igmhub/picca/blob/0f2b7b19b75f9ef61db3d180bb1c535362f7e8f0/py/picca/tests/delta_extraction/expected_flux_tests.py#L393). In that particular case I'd probably also set the branching differently so the decision if the function needs calling is made within `survey.py`, while the function itself just selects the best exp. I also think that the check should not be based upon checking if a unique target list is the same as the whole target list, but instead one should carry around a variable/config option to see if those things are needed... That would clarify why things happen in a certain way without cluttering for those uniqueness tests.

Waelthus commented:
I added a skeleton for the data readin from tiles. Lets see how the coverage reacts

Waelthus commented:
Ok, testing multiple ways of the tile based non-coadded analysis (i.e. with/without keeping individual spectra and one vs multiple CPUs) actually makes the coverage test pass, nevertheless would be good to have a test for the mean expected flux and healpix mode.

Waelthus commented on bin/picca_Pk1D.py:
```diff
@@ -272,7 +235,7 @@ def process_all_files(index_file_args):
                                  ((k < args.kmax_noise_avg) if args.kmax_noise_avg is not None else 1))
                 mean_pk_noise = np.mean(pk_noise[selection])
-                pk = (pk_raw - pk_noise) / correction_reso
+                pk = (pk_raw - mean_pk_noise) / correction_reso
```
this is now switching the defaults from taking the bin wise noise power to taking the mean, so would be good to have that documented (except if we've used the mean for all previous analyses, just not as the default here)

Waelthus commented on py/picca/delta_extraction/data.py:
```diff
@@ -600,3 +645,65 @@ def save_deltas(self):
+        return self.forests
+
+
+    def return_coadded_forests(self):
```
this function is currently not called anywhere

Waelthus commented on py/picca/delta_extraction/data.py:
```diff
@@ -600,3 +645,65 @@ def save_deltas(self):
+                for i, forest in enumerate(
+                    np.array(self.forests)[los_id_list == los_id]
+                ):
+                    forest.los_id = f"{los_id}_{i}"
```
still not sure if this los_id hack for combining an exposure index with the los_id is a good thing to do (would e.g. actually prefer this to be two different properties and to not just give a random index i, but having the EXPID encoded in there). But should be good enough for the moment.

Waelthus commented on py/picca/delta_extraction/data.py:
```diff
@@ -600,3 +645,65 @@ def save_deltas(self):
+                    forest.los_id = f"{los_id}_{i}"
+
+
+    def select_best_exposures(self):
```
this is currently called on all analyses by survey.py when computing mean expected flux, I think the decision on if we need to select the best exposure should go there

corentinravoux replied:
in the normal case, it does not select the best exposure.

Waelthus replied:
yes, but I think it would be nicer to have that function actually just do the selection, and only call it from survey.py when we want to select a best exposure...

Waelthus commented on py/picca/tests/delta_extraction/data_tests.py:
```diff
@@ -1315,6 +1316,86 @@ def test_desi_tile_read_data(self):
+
+        data = DesiTile(config["data"])
+
+        self.assertTrue(len(data.forests) == 110)
```
could check more than just the length of forests, here, if useful

Waelthus commented on py/picca/tests/delta_extraction/data_tests.py:
```diff
@@ -1315,6 +1316,86 @@ def test_desi_tile_read_data(self):
+
+        data = DesiTile(config["data"])
+
+        self.assertTrue(len(data.forests) == 10)
```
not sure if it's useful to have tests of the coadding within picca in Pk1d mode (this test and the one after), probably takes a long time while testing, potentially remove again, @iprafols?

corentinravoux commented:
I am currently testing on iron: I made one run which worked with best exposure CF, and I am running separated exposure CF to compare.
Once we decide the best way to perform CF, we can implement the test on delta extraction, and maybe finally close this PR.

corentinravoux commented:
Copying slack message without plots, as it is private data:

I have computed P1D on Y1 with two ways to measure the CF:
CF on the best exposure, applied to all exposures
CF on all exposures independently
The result for independent exposures is almost perfect, no deviation to Y1 in the statistical errors bars of Y1 which are really low. And this without any modeling of the noise. With this result, I am starting wondering if it is necessary to look for correlated noise sources.
The best exposure case result is not that good. There seems to be an outlier in the z=2.4 bin. But even without this one, there is a 3/4% shift for almost all scales. To understand this shift I checked the continuum of a specific quasar which have 2 independent exposures. The continuum is shown on the third plot for each exposure and for the coadd (realized in the previous runs of Y1). It seems like computing the delta of an exposure with the continuum computed on another one (either the best SNR exposure but I think also the coadd) will bias the deltas and consequently the P1D.
Here, the stack of delta is not forced to zero, this solution could also help to reduce this bias, but my main conclusion is that treating exposures independently is enough.

corentinravoux commented:
Since the independent exposure continuum fitting case works very well, I think I will let this solution for the PR, and remove the others solutions.
@Waelthus What do you think about that ?

Waelthus commented:
I think given the biases the "best exp" case might not be a good one to keep and having the 2 options of either using individual exposures for cont fitting (thus neglecting that the true QSO continuum should be ~ the same (although when there's long timelags between the obs it can also be the quasar changed a bit)) and co-adding (which I believe you've sped up). 
Having 2 alternative options here should be enough to allow reproducing that test later on, e.g. when writing the paper or dealing with ref comments... Agreed that the main analysis should probably be run on the indiv exp case if that's the only relatively unbiased estimate...
Did you add options for selecting different exp only if they are done on different tiles as well etc? Such that teff is basically nominal for all exps (instead of having e.g. split exps as different exposures, or aren't they?)

corentinravoux commented:
Ok, so for the last step of this PR, I will try to had an option which defines how the CF is performed, and it will also allows to separete the standard case to the cross-exposure case.
For your last question, this is an interesting point. For now I am working on the healpix, and after checking on the spectra directory, a very large portion of the exposures are on the same tile (which is weird by the way). So, I do not think selecting on this criteria would be optimal.

corentinravoux commented:
After consultation with the desi data team, the best solution to get rid of both splitted exposures and correlated noise might be to keep only one repetition of TARGETID+NIGHT. I do not know if we add it to this PR.

Waelthus commented:
What does this mean in here? Taking only one exp or rather taking a coadd of all same-TARGETID exposures in the same night, but not coadding across nights?

Waelthus commented:
I guess it could be ok to only take the best of the split exposures in that case (I guess typical splits have most of the SNR in the first spectrum [conditions worsen while observing leading to an early stopping] OR have about equal T_eff when conditions were not great, but good enough for splits), but that would have less SNR than usual... Depending on how often exposures are actually split one could also think about removing splits entirely (that would keep the sample uniform, but we'd lose more spectra)...

corentinravoux commented:
With this new commit, I have put all the options we wanted.
We can focus on the testing and merge now.

corentinravoux commented:
Hi @Waelthus, Thanks a lot for all those fix !
I am happy with the state of the PR, do you still have in mind unresolved issues or potential options to implement ?

Waelthus commented on py/picca/delta_extraction/data.py:
```diff
@@ -37,6 +37,7 @@
     "rejection log file",
     "save format",
     "num processors",
+    "delta extraction single exposure",
```
not so sure this name is really clear, potentially should add something to the parameter descriptions in the readme-files or have a more self-explanatory name. Maybe something like "spectral coadding method", but given that this is only applied for cont fitting might be misleading... @corentinravoux @iprafols any ideas?

iprafols replied:
I think that as long as we explain this somewhere this name should be fine

Waelthus reviewed (approved):
ok, I had a quick last look and think this is ok. If @iprafols is fine as well we could merge. 
Basic tests are included, but end-to-end is still missing afaik. Should keep this in mind for future development.
Potentially should add the new parameters etc to the relevant drawings etc...

iprafols commented:
I am fine with merging
